swesmith/profiles/base.py

```diff
@@ -0,0 +1,314 @@
+
+    This class provides a language-agnostic interface for repository configuration,
+    allowing different languages (Python, Go, Rust, etc.) to have their own
+    installation and testing patterns while maintaining a consistent API.
+    """
+
+    owner: str
+    repo: str
+    commit: str
+    org_dh: str = ORG_NAME_DH
+    org_gh: str = ORG_NAME_GH
+    arch = "x86_64" if platform.machine() not in {"aarch64", "arm64"} else "arm64"
+    pltf = "linux/x86_64" if arch == "x86_64" else "linux/arm64/v8"
+
+    # Install + Test specifications
+    install_cmds: list[str] = None
+    test_cmd: str = None
+    test_exts: str = [".py", ".go", ".rb", ".php", ".java"]
+
+    # `min_testing`: If set, then subset of tests (not all) are run for post-bug validation
+    # Affects get_test_cmd, get_valid_report
+    min_testing: bool = False
+
+    # `min_pregold`: If set, then for pre-bug validation, individual runs are
+    # performed instead of running the entire test suite
+    # Affects valid.py
+    min_pregold: bool = False
+
+    def __init__(self):
+        """Initialize the repository profile with an instance-level lock."""
+        # NOTE on this design: the lock is to prevent concurrent clones of the same repository.
+        # In this repo, all subclasses of RepoProfile are meant to be Singletons (only one instance
+        # of the class will ever be created). If this changes for some reason in the future,
+        # this design may have to be updated.
+        self._lock = Lock()
+
+    @abstractmethod
+    def build_image(self):
+        """Build a Docker image (execution environment) for this repository profile."""
+        pass
+
+    @abstractmethod
+    def log_parser(self, log: str) -> dict[str, str]:
+        """Parse test output logs and extract relevant information."""
+        pass
+
+    @property
+    def image_name(self) -> str:
+        return f"{self.org_dh}/swesmith.{self.arch}.{self.owner}_1776_{self.repo}.{self.commit[:8]}".lower()
+
+    @property
+    def mirror_name(self):
+        return f"{self.org_gh}/{self.repo_name}"
+
+    @property
+    def repo_name(self):
+        return f"{self.owner}__{self.repo}.{self.commit[:8]}"
+
+    def _mirror_exists(self):
+        """Check if mirror repository exists under organization"""
+        return self.repo_name in [
+            x["name"]
+            for page in range(1, 3)  # TODO: Need to update over time
+            for x in api.repos.list_for_org(self.org_gh, per_page=100, page=page)
+        ]
+
+    def clone(self, dest: str | None = None) -> str | None:
+        """Clone repository locally"""
+        if not self._mirror_exists():
+            raise ValueError(
+                "Mirror clone repo must be created first (call .create_mirror)"
+            )
+        dest = self.repo_name if not dest else dest
+        if not os.path.exists(dest):
+            clone_cmd = (
+                f"git clone git@github.com:{self.mirror_name}.git"
```

Comment:
slightly faster way if we really don't need the history:

```
                            "git init",
                            f"git remote add origin {url}",
                            f"git fetch --depth 1 origin {base_commit}",
                            "git checkout FETCH_HEAD",
```

This will not try to get the whole history, which can take minutes in very large repos and requires lots of space